src/sbuf.c

```diff
@@ -560,16 +677,98 @@ static bool sbuf_send_pending(SBuf *sbuf)
 	goto try_more;
 }
 
+/*
+ * There's data in extra_packets buffer to be sent. Returns bool if processing
+ * can continue.
+ */
+static bool sbuf_send_pending_extra_packets(SBuf *sbuf)
+{
+	int avail;
+	ssize_t res;
+	struct MBuf *mbuf = &sbuf->extra_packets;
+
+	AssertActive(sbuf);
+	Assert(sbuf->dst || mbuf_avail_for_read(mbuf) == 0);
+	log_noise("sbuf_send_pending_extra_packets ");
+
+try_more:
+	/* how much data is available for sending */
+	avail = mbuf_avail_for_read(mbuf);
+	if (avail == 0)
+		return true;
+
+	if (sbuf->dst->sock == 0) {
+		log_error("sbuf_send_pending_extra_packets: no dst sock?");
+		sbuf_call_proto(sbuf, SBUF_EV_SEND_FAILED);
+		return false;
+	}
+
+	/* actually send it */
+	//res = iobuf_send_pending(io, sbuf->dst->sock);
+	res = sbuf_op_send(sbuf->dst, mbuf->data + mbuf->read_pos, avail);
+	if (res > 0) {
+		mbuf->read_pos += res;
+	} else if (res < 0) {
+		if (errno == EAGAIN) {
+			if (!sbuf_queue_send(sbuf)) {
+				/* drop if queue failed */
+				sbuf_call_proto(sbuf, SBUF_EV_SEND_FAILED);
+			}
+		} else {
+			sbuf_call_proto(sbuf, SBUF_EV_SEND_FAILED);
+		}
+		return false;
+	}
+
+	AssertActive(sbuf);
+
+	/*
+	 * Should do sbuf_queue_send() immediately?
+	 *
+	 * To be sure, let's run into EAGAIN.
+	 */
+	goto try_more;
+}
+
+
 /* process as much data as possible */
 static bool sbuf_process_pending(SBuf *sbuf)
 {
 	unsigned avail;
 	IOBuf *io = sbuf->io;
+	struct MBuf *mbuf = &sbuf->extra_packets;
 	bool full = iobuf_amount_recv(io) <= 0;
-	bool res;
+	int loop_number = 0;
+	log_noise("sbuf_process_pending: start");
 
 	while (1) {
+		/*
+		 * If there's still queued extra packets from a previous packet, make
+		 * sure to flush those first. Otherwise the packet we process next
+		 * might add even more packets there, which would be bad because it
+		 * would mean they get delivered out of order.
+		 */
+		if (mbuf_avail_for_read(mbuf)) {
+			if (sbuf->extra_packet_queue_after) {
+				if (!sbuf_send_pending_iobuf(sbuf)) {
+					log_noise("sbuf_process_pending failed to send all pending data");
+					return false;
+				}
+			}
+
+			if (!sbuf_send_pending_extra_packets(sbuf)) {
+				log_noise("sbuf_process_pending ended early because of not being able to send the queued extra packets");
+				return false;
+			}
+			/*
+			 * XXX: Should we maybe free this if it is too large?
```

Comment:
Let's decide on this. I think the current behaviour of never freeing is probably undesired. If a client prepares a big query it will bloat memory usage of the client for the lifetime of the client. I think it would be best to free the buffer if it grew above n-times the value of `pkt_buf` (where n is probably 1 or 2).

(cc @turnstep since you caused me to write about the current memory behaviour in the README)

Reply:
Yeah, at least n=2, maybe slightly more. Perhaps configurable? ORMs like to list and quote every single column, leading to all of their queries being quite large.

Reply:
Made it 4x pkt_buf which still seems reasonable. I think it's unlikely that people will manually need to configure this and if they do they can simply increase pkt_buf. So I'd rather not add yet another configuration option to pgbouncer (there's already a lot of them).